### PR TITLE
ghc: fix build on linux

### PR DESCRIPTION
**Pull request by kyungminlee**
Fixes https://github.com/Linuxbrew/homebrew-core/issues/1856.

### CODE REVIEW

File: Formula/ghc.rb [hunk 1]
```diff
@@ -122,7 +123,7 @@ def install
 
     if OS.linux?
       # Fix error while loading shared libraries: libgmp.so.3
```
Comment (rwhogg) [inline Formula/ghc.rb]: Please also update the comment.

**Comment by rwhogg** [comment]
From https://travis-ci.org/Linuxbrew/homebrew-core/builds/204842850#L1983
```
/tmp/ghc-20170224-22074-zxtapm/ghc-8.0.2/binary/lib/ghc-7.8.4/bin/ghc-pkg: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
```
libtinfo.so.5 is provided by the Ubuntu package `libtinfo5`. `libtinfo5` is installed, but `libtinfo-dev` is removed by Travis. That *might* be the cause of the problem.

Possibly also relevant is that the `ncurses` formula includes a libtinfo.so (not libtinfo.so.5) symlink, but not a tinfo.pc. @sjackman, on https://github.com/Homebrew/homebrew-dupes/pull/517#issuecomment-222375634 you mentioned that you couldn't get `--with-termlib` to work; maybe we should revisit that?

**Comment by sjackman** [comment]
I don't think we want to use `--with-termlib`. This is a weird case base we're trying to run an executable binary `ghc-pkg` that was not compiled by Linuxbrew, and has deps on libraries not provided by Linuxbrew.

File: Formula/ghc.rb [hunk 2]
```diff
@@ -121,8 +122,8 @@ def install
     end
 
     if OS.linux?
-      # Fix error while loading shared libraries: libgmp.so.3
-      ln_s Formula["gmp"].lib/"libgmp.so", gmp/"lib/libgmp.so.3"
+      # Fix error while loading shared libraries: libgmp.so.10
+      ln_s Formula["gmp"].lib/"libgmp.so", gmp/"lib/libgmp.so.10"
```
Comment (sjackman) [inline Formula/ghc.rb]: Try
```ruby
ln_s Formula["ncurses"].lib/"libtinfo.so", gmp/"lib/libtinfo.so.5"
```

File: Formula/ghc.rb [hunk 3]
```diff
@@ -7,6 +7,7 @@ class Ghc < Formula
   homepage "https://haskell.org/ghc/"
   url "https://downloads.haskell.org/~ghc/8.0.2/ghc-8.0.2-src.tar.xz"
   sha256 "11625453e1d0686b3fa6739988f70ecac836cadc30b9f0c8b49ef9091d6118b1"
+  revision 1
```
Comment (sjackman) [inline Formula/ghc.rb]: Please revert this change before committing.

File: Formula/ghc.rb (hunk 2 above)
Comment (kyungminlee) [inline Formula/ghc.rb]: The symbolic link `libtinfo.so.5` is confusing `/bin/bash`...

```
./configure
--prefix=/tmp/ghc-20170226-22842-iojoqx/ghc-8.0.2/binary
--with-gmp-includes=/home/linuxbrew/.linuxbrew/Cellar/ghc/8.0.2/libexec/integer-gmp/include
--with-gmp-libraries=/home/linuxbrew/.linuxbrew/Cellar/ghc/8.0.2/libexec/integer-gmp/lib
--with-ld=ld
--with-gcc=gcc

checking for path to top of build tree... /bin/bash: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.16' not found (required by /home/linuxbrew/.linuxbrew/Cellar/ghc/8.0.2/libexec/integer-gmp/lib/libtinfo.so.5)
configure: error: cannot determine current directory
```

Comment (kyungminlee) [inline Formula/ghc.rb]: A quick and dirty fix like https://github.com/kyungminlee/homebrew-kyungminlee/commit/436203d53a63449a41efc91016c37e49d3e224b0 (`patchelf`ing binaries and shared libraries in the `resource('binary')` with `--replace-needed`) solves the dependency problem on `libtinfo.so.5`. What do you think?

Comment (sjackman) [inline Formula/ghc.rb]: Good idea!
```ruby
safe_system "patchelf", "--replace-needed", "libtinfo.so.5", "libtinfo.so", file
```
